_drafts/assert.md

```diff
@@ -0,0 +1,662 @@
+- Good placement of asserts can reduce chances of exceptions by catching
+  out-of-bounds accesses, invalid pointer dereferences, invalid state machine
+  transitions, and nonsensical operations (like a malloc of 12MB on a 64kB MCU).
+- Asserts can help developers quickly understand assumptions made by a certain
+  piece of code (e.g. a module that takes a `uint8_t` but requires a value
+  1-64).
+
+## Common Usages of Asserts
+
+Many assertions are application-specific, but we've got a few rules of thumbs on
+when you should use assertions. Below are some examples of when using assertions
+is a good idea.
+
+### When malloc shouldn't fail
+
+Dynamic memory allocation is common in embedded systems, especially in today's
+Internet connected consumer electronics. These devices need to be able to
+receive an influx of packets, use large HTTP buffers, and possibly load assets
+or fonts into RAM for a user interface. Keeping around statically allocated
+buffers for all of these functions would be wasteful, especially for the devices
+that measure RAM in kilobytes. In these devices, dynamic memory allocation is a
+must.
+
+There are also places where malloc could fail, and where **it shouldn't**.
+
+For example, let's give an example of dynamically allocated mutexes where malloc
+should not fail.
+
+```c
+Mutex *mutex_create(void) {
+  Mutex *mutex = malloc(sizeof(Mutex));
+  ASSERT(mutex);
+  mutex_init(mutex);
+
+  return mutex;
+}
+```
+
+This malloc **should never fail**. If it did, then it would mean the system is
+out of heap space. If we didn't assert here, then the system would risk using an
+unprotected resource.
+
+On the contrary, there are cases where it is perfectly fine for malloc to fail,
+such as when a surge of packets are received and the heap becomes exhausted. The
+system should process the packets, free the memory, and then ask for more data.
+
+### Invalid arguments passed into function
+
+Argument checking is one of the best (and easiest) places to assert.
+
+We can check for NULL pointers
+
+```c
+void handle_event(Event *evt) {
+  ASSERT(evt);
+  ...
+}
+```
+
+or check sizes of buffers
+
+```c
+void read_packet(void *buf, size_t buf_len) {
+  ASSERT(buf_len >= sizeof(Packet));
+  memcpy(...)
+}
+
+```
+
+or prevent silly operations.
+
+```c
+void *my_malloc(uint32_t n) {
+  ASSERT(n < 131072);
+  ...
+}
+```
+
+All of the above examples above are **bugs** introduced by a developer and they
+likely shouldn't be handled gracefully. The alternative is to return an error
+code, (hopefully not `-1`), and burden the develaoper with finding out _exactly_
+where in the call stack a particular argument check filed.
+
+> It's not common to see asserts on arguments in vendor and library code, since
+> they aren't in the business of crashing devices, but it is sometimes
+> configurable on mature software. One example of configurable asserting in
+> library code is the `xTaskCreateStatic` function in FreeRTOS[^4].
```

Comment:
maybe "can be found in FreeRTOS." and in the annotation say i.e see `xTaskCreateStatic`

That pattern is all over the codebase but this reads to me like its only in one function